torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1652,6 +1683,16 @@ struct to_ir {
 
       auto unified = unifyTypes(tv->type(), fv->type());
 
+      // If the variable we're looking at is known to be Union[T1, T2],
+      // then it's okay to have one branch return T1 and the other
+      // return T2. (We're not actually going to use `unified` again;
+      // it's acting as a flag)
+      c10::optional<TypePtr> full_true_type = environment_stack->getType(x);
+      c10::optional<TypePtr> full_false_type = environment_stack->getType(x);
+      if (full_true_type && full_false_type) {
```

Comment:
Can you explain why `unifyTypes(tv->type(), fv->type());` call above does not handle whatever case this has been added to cover ?